### PR TITLE
Remove nondeterminism in multiple-definitions test

### LINKED ISSUE #87084: Unrelated changes make multiple-definitions test fail spuriously
The test was originally added in #86419, see the comments under that for more discussion.

#86922 and #86857 fails with this diff:
```diff
10 error: multiple definitions of external function `f` from library `foo.dll` have different calling conventions
-   --> $DIR/multiple-definitions.rs:8:5
12    |
12    |
- LL |     fn f(x: i32);
-    |     ^^^^^^^^^^^^^
+ LL |         fn f(x: i32);
```
#87075 was opened, removing the `^^^` and adding indentation (I did not see that it suggested removing `--> $DIR/...`). It fails with this diff:

```diff
10 error: multiple definitions of external function `f` from library `foo.dll` have different calling conventions
12    |
12    |
- LL |         fn f(x: i32);
+ LL |     fn f(x: i32);
14 
15 error: aborting due to previous error; 1 warning emitted
```

I just want to say that I don't think there was some nondeterministic code causing the span to be different (at least in the file that emitted the diagnostic), as shown on these lines, where they get sorted, deterministically:
https://github.com/rust-lang/rust/blob/a867dd4c7e65beb1453fcff9e252e7ab19f0e1d0/compiler/rustc_codegen_ssa/src/back/link.rs#L483-L487

@rustbot label A-spurious C-bug T-compiler

### PR DESCRIPTION
**Pull request by ricobbe**
Compare all fields in `DllImport` when sorting to avoid nondeterminism in the error for multiple inconsistent definitions of an extern function.  Restore the multiple-definitions test.

Resolves #87084.

### CODE REVIEW

**Comment by ricobbe** [comment]
Question for reviewers: in an earlier version of this code, a reviewer asked me not to have `DllImport` implement `PartialOrd` or `Ord`, I think because the ordering relation as implemented then didn't consider all of `DllImport`'s fields, but it's been a while and I don't remember for certain.  The new ordering relation I'm defining here does indeed consider all of the fields, and it's effectively equivalent to the one that we'd get if I just derived `PartialOrd` and `Ord`.  Would you prefer me to just do that instead?

**Comment by fee1-dead** [comment]
I don't think it is nondeterministic when you compared the names, so I don't think the issue is that simple.

**Comment by ricobbe** [comment]
> I don't think it is nondeterministic when you compared the names, so I don't think the issue is that simple.

The error message is triggered in precisely the case where we have multiple `DllImport`s with the same name (plus some other conditions that aren't important here).  Therefore, to get a unique, well-defined result from sorting the imports (and therefore selecting the second entry with a duplicate name for the error message), I need to compare based on more than just the names.

**Comment by petrochenkov** [comment]
My version of the fix, with all unnecessary impl removed, and with explicit sorting removed too - https://github.com/petrochenkov/rust/commit/8b9ce0e7ddf243b8c79022848c1ac85aea9ee900

**Comment by ricobbe** [comment]
> My version of the fix, with all unnecessary impls removed, and with explicit sorting removed too - [petrochenkov@8b9ce0e](https://github.com/petrochenkov/rust/commit/8b9ce0e7ddf243b8c79022848c1ac85aea9ee900)

I'm happy to just go with that; what's the workflow to do that?

**Comment by petrochenkov** [comment]
@ricobbe
You can cherry-pick that commit and add/update the test.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -471,16 +472,20 @@ fn collate_raw_dylibs(
         }
     }
 
-    // Rustc already signals an error if we have two imports with the same name but different
-    // calling conventions (or function signatures), so we don't have pay attention to those
-    // when ordering.
-    // FIXME: when we add support for ordinals, figure out if we need to do anything if we
-    // have two DllImport values with the same name but different ordinals.
+    fn compare_imports(x: &DllImport, y: &DllImport) -> Ordering {
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:475]: You can destructure to ensure that you matched all fields. Eg
```suggestion
    fn compare_imports(DllImport { x_name, x_ordinal, x_calling_convention, x_span }: &DllImport, DllImport { y_name, y_ordinal, y_calling_convention, y_span }: &DllImport) -> Ordering {
```
or
```suggestion
    fn compare_imports(x: &DllImport, y: &DllImport) -> Ordering {
        let DllImport { x_name, x_ordinal, x_calling_convention, x_span } = x;
        let DllImport { y_name, y_ordinal, y_calling_convention, y_span } = y;
```

**Comment by ricobbe** [comment]
@petrochenkov 

> You can cherry-pick that commit and add/update the test.

Ok, I've tried that, and now the restored (and renamed) multiple-declarations.rs test is failing.  I've run it with `--bless` and updated the expected error message (in the comment in the .rs file) to adjust for the change in wording, but the test still fails: 

```
error: process did not return an error status
status: exit code: 0
```

Everything else looks reasonable; in particular, I'm seeing the expected diagnostics on stderr.

I've tried running the test with the `// check-pass` header command, and also with `// build-pass`, but neither have any effect on the results.  This is not a failure mode I know how to debug, I'm afraid.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 1 above)
Comment (ricobbe) [reply compiler/rustc_codegen_ssa/src/back/link.rs]: Made moot by petrochenkov's rewrite.

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 2]
```diff
@@ -454,65 +461,46 @@ fn link_rlib<'a, B: ArchiveBuilder<'a>>(
...
+    if error_detected {
+        Err(ErrorReported)
+    } else {
+        Ok(dylib_table
+            .into_iter()
+            .map(|(name, imports)| {
+                (name, imports.into_iter().map(|(_, import)| import.clone()).collect())
+            })
+            .collect())
+    }
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/link.rs:503]:
```suggestion
    sess.compile_status()?;
    Ok(dylib_table
        .into_iter()
        .map(|(name, imports)| {
            (name, imports.into_iter().map(|(_, import)| import.clone()).collect())
        })
        .collect())
```

**Comment by petrochenkov** [comment]
r=me after addressing https://github.com/rust-lang/rust/pull/87092#discussion_r671163114 and squashing commits.

**Pull request merged**